core/src/main/scala/kafka/log/Log.scala

```diff
@@ -181,32 +188,46 @@ class Log(val dir: File,
         // if its a log file, load the corresponding log segment
         val start = filename.substring(0, filename.length - LogFileSuffix.length).toLong
         val indexFile = Log.indexFilename(dir, start)
+        val timeIndexFile = Log.timeIndexFilename(dir, start)
         val segment = new LogSegment(dir = dir,
                                      startOffset = start,
                                      indexIntervalBytes = config.indexInterval,
                                      maxIndexSize = config.maxIndexSize,
                                      rollJitterMs = config.randomSegmentJitter,
                                      time = time,
                                      fileAlreadyExists = true)
-
-        if(indexFile.exists()) {
+        if (indexFile.exists() && timeIndexFile.exists()) {
           try {
-              segment.index.sanityCheck()
+            segment.index.sanityCheck()
+            segment.timeIndex.sanityCheck()
           } catch {
             case e: java.lang.IllegalArgumentException =>
               warn("Found a corrupted index file, %s, deleting and rebuilding index...".format(indexFile.getAbsolutePath))
               indexFile.delete()
+              timeIndexFile.delete()
               segment.recover(config.maxMessageSize)
           }
-        }
-        else {
+        } else {
           error("Could not find index file corresponding to log file %s, rebuilding index...".format(segment.log.file.getAbsolutePath))
           segment.recover(config.maxMessageSize)
         }
         segments.put(start, segment)
       }
     }
 
+    /**
+     * We append a time index entry for non-active segment with an empty time index. This could happen during upgrade.
+     * We need to do this because searching by timestamp needs a time index entry.
+     */
+    val logSegmentsSeq = logSegments.toSeq
+    for (i <- 0 until logSegmentsSeq.size - 1) {
+      // load the largest timestamp into memory.
+      logSegmentsSeq(i).loadLargestTimestamp()
+      // Insert the last entry to the time index for inactive segments.
+      if (logSegmentsSeq(i) != activeSegment)
+        logSegmentsSeq(i).maybeAppendLastTimeIndexEntry(logSegmentsSeq(i + 1).baseOffset)
```

Comment:
Hmm, it seems that we should insert the offset in segment i that has the largest timestamp. Not sure why we want to insert the offset in the next segment.